src/liballoc/heap.rs

```diff
@@ -44,6 +45,84 @@ fn check_size_and_alignment(size: usize, align: usize) {
                   align);
 }
 
+#[derive(Copy, Clone, Default, Debug)]
+pub struct HeapAllocator;
+
+unsafe impl Allocator for HeapAllocator {
+    unsafe fn alloc(&mut self, layout: Layout) -> Result<*mut u8, AllocErr> {
```

Comment:
I've removed the `_unchecked` variants, and added explicit conditions to the core `Allocator` methods saying that behavior is undefined for zero-sized layouts.

My personal opinion is that it was a mistake to have a debug_assert! on zero size input in the underlying global allocator methods, but its there for now and I would rather make the `Allocator` API compatible with them.

if we can get buy-in to make the global allocators' behavior closer to standard C (which specifies that you get either a null pointer or a nonusable non-null pointer), then we can revise the `Allocator` specification later to reflect that narrowing of the specification.